accord-core/src/main/java/accord/local/CommandSummaries.java

```diff
@@ -60,185 +62,279 @@ enum SummaryStatus
-                return factory.create(primaryTxnId, keysOrRanges, redundantBefore, kinds, minTxnId, maxTxnId, findAsDep);
-            }
+        public static <L extends SummaryLoader, P> L loader(P param, RedundantBefore redundantBefore, MaxDecidedRX maxDecidedRX, TxnId primaryTxnId, LoadKeysFor loadKeysFor, Unseekables<?> keysOrRanges, Factory<L, P> factory)
+        {
+            Invariants.require(primaryTxnId != null);
+            TxnId minTxnId = redundantBefore.min(keysOrRanges, Bounds::gcBefore);
+            Timestamp maxTxnId = loadKeysFor == RECOVERY || !primaryTxnId.is(ExclusiveSyncPoint) ? Timestamp.MAX : primaryTxnId;
+            TxnId findAsDep = loadKeysFor == RECOVERY ? primaryTxnId : null;
+            Kinds kinds = primaryTxnId.witnesses().or(loadKeysFor == RECOVERY ? primaryTxnId.witnessedBy() : Nothing);
+            if (!primaryTxnId.is(Txn.Kind.ExclusiveSyncPoint))
+                maxDecidedRX = null;
+            return factory.create(param, redundantBefore, maxDecidedRX, primaryTxnId, keysOrRanges, kinds, minTxnId, maxTxnId, findAsDep);
+        }
 
-            public Loader(@Nullable TxnId primaryTxnId, Unseekables<?> searchKeysOrRanges, RedundantBefore redundantBefore, Kinds testKind, TxnId minTxnId, Timestamp maxTxnId, @Nullable TxnId findAsDep)
-            {
-                this.searchKeysOrRanges = searchKeysOrRanges;
-                this.redundantBefore = redundantBefore;
-                this.testKind = testKind;
-                this.minTxnId = minTxnId;
-                this.maxTxnId = maxTxnId;
-                this.findAsDep = findAsDep;
-            }
+        public SummaryLoader(Object ignore, RedundantBefore redundantBefore, MaxDecidedRX maxDecidedRX, TxnId primaryTxnId, Unseekables<?> searchKeysOrRanges, Kinds testKind, TxnId minTxnId, Timestamp maxTxnId, @Nullable TxnId findAsDep)
+        {
+            this.redundantBefore = redundantBefore;
+            this.maxDecidedRX = maxDecidedRX;
+            this.primaryTxnId = primaryTxnId;
+            this.searchKeysOrRanges = searchKeysOrRanges;
+            this.testKind = testKind;
+            this.minTxnId = minTxnId;
+            this.maxTxnId = maxTxnId;
+            this.findAsDep = findAsDep;
+            this.minDecidedId = minDecidedDependencyId(maxDecidedRX, searchKeysOrRanges, primaryTxnId);
+        }
 
-            public Summary ifRelevant(Command cmd)
-            {
-                return ifRelevant(cmd.txnId(), cmd.executeAtOrTxnId(), cmd.saveStatus(), cmd.participants(), cmd.partialDeps());
-            }
+        public boolean isRelevant(CommandsForKey cfk)
+        {
+            if (cfk == null || cfk.size() == 0)
+                return false;
 
-            private boolean isEligibleDep(SummaryStatus status, TxnId findAsDep, TxnId txnId, Timestamp executeAt)
+            // NOTE: we CANNOT safely filter on first element, as we may have pruned dependencies we need to witness
```

Comment:
It is not clear to me here what you mean by first element; also because we then filter by _last_

Reply:
I mean we are filtering based on the last element in the CFK, and while it might seem like we could do equivalent filtering based on the first/min element and maxTxnId, we can't.